pyvera/subscribe.py

```diff
@@ -123,23 +123,33 @@ def _run_poll_server(self):
                 device_data, timestamp = (
                     controller.get_changed_devices(timestamp))
                 LOG.debug("Poll returned")
+
                 if self._exiting:
                     continue
-                if not device_data:
+
+                if device_data:
+                    self._event(device_data)
+                else:
                     LOG.debug("No changes in poll interval")
-                    continue
-                self._event(device_data)
-                time.sleep(1)
-            except requests.RequestException:
-                LOG.info("Could not contact Vera - will retry in %ss",
-                         SUBSCRIPTION_RETRY)
-                time.sleep(SUBSCRIPTION_RETRY)
-            except json.decoder.JSONDecodeError:
-                LOG.info("Response was malfomed - will retry in %ss",
-                         SUBSCRIPTION_RETRY)
-                time.sleep(SUBSCRIPTION_RETRY)
+
+                continue
+
+            except requests.RequestException as ex:
+                LOG.debug("Caught RequestException: %s", str(ex))
+                pass
+
+            except json.decoder.JSONDecodeError as ex:
+                LOG.debug("Caught JSONDecodeError: %s", str(ex))
+                pass
+
             except Exception as ex:
-                LOG.exception("Vera thread exception %s", ex)
-                raise
+                LOG.exception("Vera poll thread general exception: %s", 
+                              str(ex))
+                # Don't raise. Stay in control
+                #raise
```

Comment:
Let's make a decision - if we're going to catch everything - delete the commented out raise.

My intent with the trace was to add additional exceptions as they came up. But I can see your point above.

Reply:
OK. Here's my vote. I think in doing the JSON decode, that ONE statement in its own try block, we should trap anything that comes at up, and report it back as a parsing error in some form, either abusing an existing exception as I did, or introducing a new one. In the loop itself, we can remove the blanket exception, or more to my preference, keep it, message it, but uncomment the raise and let it continue to propagate. If we want to log more in those areas, I'm never opposed to that. I'm also in the "more information is better than less" camp.

WRT the callbacks, I think we should do the same. A callback should be trapping and reporting its own exceptions. If it gets to us, it's serious heartburn, but we should not let that take out the entire thread. On my Vera, there are 150 devices, and one device in particular occasionally gives an HA callback heartburn and knocks the whole thing out: it returns a string longer than 255 characters, and the HA callback chokes somewhere on it. That callback is making two bad choices, one of which is not trapping its own errors. I think we should blanket-catch that ONE statement (the invocation of the callback), log at error level the untrapped (now pyvera-trapped) exception, and continue on.

This gives those two cases VERY limited scope, with a reasonable expectation of good behavior on the part of the parser, and good self-management on the part of the callback. 

If this was my code, I would actually refactor the entire polling loop so that the get_changed call was in its own try small block (to catch whatever is passed on by that call that should not kill the thread--just enough to distinguish temporary errors from likely permanent ones), and the callback invocation was in its own one-statement try block, but everything else was naked, because really, there's nothing else that should be throwing elsewhere in that code, other than latent logic/coding errors or odd exceptions for which you would want a good, untrapped exception to be thrown (with its very accurate traceback associated).

That's my 2p. I'm open to whatever fits with your sensibilities. I am, of course, just trying to make the code harder to tip over.

Reply:
OK. I dug in a little more, trying to narrow the scope of the polling loop checks as much as possible. From what I've been able to discover, even the loosest of the python JSON parsers likely to be used derives their exceptions from ValueError, so restricting the scope of the parsing catches to that class seems like an acceptable, desirable middleground. I'm testing that now, beating the heck out of my Vera every way I can think of. So far, so good.